applications/PfemFluidDynamicsApplication/custom_processes/set_material_properties_for_thermal_coupling_process.hpp

```diff
@@ -0,0 +1,174 @@
+#define  SET_MATERIAL_PROPERTIES_FOR_THERMAL_COUPLING_PROCESS
+
+#include "includes/variables.h"
+#include "utilities/variable_utils.h"
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "processes/process.h"
+#include "includes/convection_diffusion_settings.h"
+#include "includes/constitutive_law.h"
+
+namespace Kratos
+{
+
+  /**
+   * @class SetMaterialPropertiesForThermalCouplingProcess
+   * @ingroup PfemFluidDynamicsApplication
+   * @brief This process sets the nodal value of thermal properties (density, conductivity and capacity)
+   * that depends on the temperature, which is necesssary for solving the termal part,
+   * since the convection-diffusion solver gets these nodal values for assembling the system.
+   * The value of these properties are computed according to the constitutive law of the incident elements of each node,
+   * using the nodal temperature, and the average is taken as the nodal value for the property.
+   * @author Rafael Rangel
+  */
+  class SetMaterialPropertiesForThermalCouplingProcess : public Process {
+
+  public:
+
+    KRATOS_CLASS_POINTER_DEFINITION(SetMaterialPropertiesForThermalCouplingProcess);
+
+    /// Constructor
+    explicit SetMaterialPropertiesForThermalCouplingProcess(ModelPart& fluid_model_part, ModelPart& thermal_model_part) :
+      rFluidModelPart(fluid_model_part),
+      rThermalModelPart(thermal_model_part) {}
+
+    /// Destructor.
+    ~SetMaterialPropertiesForThermalCouplingProcess() override {}
+
+    void operator()() {
+      Execute();
+    }
+
+    void Execute() override {
+      KRATOS_TRY;
+
+      this->SetMaterialProperties(rFluidModelPart, rThermalModelPart);
+
+      KRATOS_CATCH("");
+    }
+
+    void ExecuteInitialize() override {}
+
+    void ExecuteInitializeSolutionStep() override {}
+
+  protected:
+
+    ModelPart& rFluidModelPart;
+    ModelPart& rThermalModelPart;
+
+  private:
+
+    void SetMaterialProperties(ModelPart& rFluidModelPart, ModelPart& rThermalModelPart) const {
+
+        const ProcessInfo& rCurrentProcessInfo = rFluidModelPart.GetProcessInfo();
+        ConvectionDiffusionSettings::Pointer my_settings = rCurrentProcessInfo.GetValue(CONVECTION_DIFFUSION_SETTINGS);
+
+        // Loop over all nodes
+        const auto& i_node_begin = rFluidModelPart.NodesBegin();
+        for (int i = 0; i < static_cast<int>(rFluidModelPart.Nodes().size()); i++) {
+          auto i_node = i_node_begin + i;
+
+          // Nodal temperature
+          double temp = i_node->FastGetSolutionStepValue(TEMPERATURE);
+
+          // Initialize thermal properties
+          double density      = 0.0;
+          double conductivity = 0.0;
+          double capacity     = 0.0;
+
+          // Loop over incident elements
+          ElementWeakPtrVectorType& neighbour_elements = i_node->GetValue(NEIGHBOUR_ELEMENTS);
+          int nel = neighbour_elements.size();
+          int n = 0;
+
+          for (auto& i_nelem : neighbour_elements) {
```

Comment:
this can be done in parallel?

Reply:
Now I am making the outside loop over all nodes, which is more expensive, in parallel. The inner loop over neighbour elements needs parallelization?